### PR TITLE
nrf/51: Initial commit to unicore-mx project. Basic support for Clocks, GPIO and 2.4GHz Radio

### PR DESCRIPTION
**Pull request by maksymko**

### CODE REVIEW

File: include/unicore-mx/nrf/51/gpio.h [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+#define GPIO_IN			MMIO32(GPIO_BASE + 0x510)
+
+#define GPIO_DIR			MMIO32(GPIO_BASE + 0x514)
+#define GPIO_DIRSET			MMIO32(GPIO_BASE + 0x518)
+#define GPIO_DIRCLR			MMIO32(GPIO_BASE + 0x51C)
+
+#define GPIO_PIN_CNF(N)			MMIO32(GPIO_BASE + 0x700 + 0x4 * (N))
+
+#define PIN0			(1 << 0)
```
Comment (danielinux) [inline include/unicore-mx/nrf/51/gpio.h]: I'd rather use `GPIOPIN0`...`GPIOPIN31` to avoid namespace clashes, and to make it more similar to NXP LPC gpio port

File: include/unicore-mx/nrf/51/gpio.h [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+void gpio_set(uint32_t gpios);
+void gpio_clear(uint32_t gpios);
+void gpio_toggle(uint32_t gpios);
+uint32_t gpio_get(uint32_t gpios);
+
+void gpio_set_direction(uint32_t gpios, enum gpio_dir dir);
+void gpio_set_pull(uint32_t gpios, enum gpio_pull pull);
+
+void gpio_configure_task(uint8_t task_num,
```
Comment (danielinux) [inline include/unicore-mx/nrf/51/gpio.h:176]: Can we reuse function prototypes from other platforms such as  gpio_mode_setup/gpio_set_output_options/gpio_set_af, and keep arguments "generic" as uint32_t? This would facilitate our task of having the GPIO API as much unified as possible. Enums above should then be replaced by #defines

**Comment by danielinux** [comment]
Hello @maksymko . Thanks for submitting this. IMHO almost good to go.
Out approach (as opposed to libopencm3) is to keep APIs as much generic as possible, and hide platforms specific symbol names. Please see my inline comments, this should be minor rework. Feel free to let us know your opinion in this regard.

**Comment by brabo** [comment]
@maksymko Hi again!
Oh neato!
I will review it very shortly, first impression is: neato! ;)

File: lib/nrf/51/Makefile [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+AR		= $(PREFIX)-ar
+TGT_CFLAGS	= -Os -Wall -Wextra -I../../../include -fno-common \
+		  -mcpu=cortex-m0 -mthumb $(FP_FLAGS) -Wstrict-prototypes \
+		  -ffunction-sections -fdata-sections -MD -DNRF51
+TGT_CFLAGS      += $(DEBUG_FLAGS)
+# ARFLAGS	= rcsv
+ARFLAGS		= rcs
+OBJS		= gpio.o rtc.o clock.o ppi.o uart.o radio.o timer.o
+
+VPATH += ../../usb:../../cm3:../common
```
Comment (brabo) [inline lib/nrf/51/Makefile]: Given there is no nrf/common directory probably we should remove ../common from VPATH :)

File: lib/nrf/51/ppi.c [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+ *
+ * This is the alternative API, which requires the caller to store the mask of used channels.
+ *
+ * @param[in/out] chan_map uint32_t* The mask of channels that are already in use.
+ *  For the first call initialize with zero and pass in.
+ * @param[in] chan_num uint8_t the number of the channel to remove from the map.
+ */
+void ppi_clear_channel(uint32_t* chan_map, uint8_t chan_num)
+{
+	ppi_disable_channels(PPI_CH(chan_num));
```
Comment (brabo) [inline lib/nrf/51/ppi.c:136]: It would be nice if we could have some consistency in function namings. We realise we need to provide a clear API definition as soon as possible. I would suggest we name it "disable" as that is very unambiguous.

File: lib/nrf/51/radio.c [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+ */
+void radio_set_crclen(uint8_t crc_len)
+{
+	uint32_t reg_crc = RADIO_CRCCNF;
+	reg_crc &= ~RADIO_CRCCNF_LEN_MASK;
+	RADIO_CRCCNF = reg_crc | RADIO_CRCCNF_LEN_MASKED(crc_len);
+}
+
+/** @brief Disable CRC calculation. */
+void radio_crc_disable(void)
```
Comment (brabo) [inline lib/nrf/51/radio.c]: Consistent naming here would be "radio_disable_crc"

File: lib/sam/4l/gpio.c [hunk 6]
```diff
@@ -60,8 +60,8 @@ void gpio_toggle(uint32_t gpioport, uint32_t gpios)
 
 /** @brief Enable output pins.
  *
- * Onlyc the ones where bits are set to "1" are touched, everything else
- * remains in the old state.
+ * @details Only the ones where bits are set to "1" are touched, everything
+ * else remains in the old state.
  *
```
Comment (brabo) [inline lib/sam/4l/gpio.c:65]: These changes seem unrelated to this PR?

**Comment by brabo** [comment]
@maksymko great work! :)
Indeed almost good to go!
:+1:

File: lib/nrf/51/ppi.c (hunk 4 above)
Comment (maksymko) [inline lib/nrf/51/ppi.c:136]: "remove" probably more apt, because it does more than just disabling it.

File: lib/sam/4l/gpio.c (hunk 6 above)
Comment (maksymko) [inline lib/sam/4l/gpio.c:65]: Removed

File: lib/nrf/51/radio.c (hunk 5 above)
Comment (maksymko) [inline lib/nrf/51/radio.c]: Fixed

File: lib/nrf/51/Makefile (hunk 3 above)
Comment (maksymko) [inline lib/nrf/51/Makefile]: Fixed

File: include/unicore-mx/nrf/51/gpio.h (hunk 1 above)
Comment (maksymko) [inline include/unicore-mx/nrf/51/gpio.h]: Fixed

File: include/unicore-mx/nrf/51/gpio.h (hunk 2 above)
Comment (maksymko) [inline include/unicore-mx/nrf/51/gpio.h:176]: I fixed the api to be generic, but reusing the API from other parts may be somewhat problematic, because this one has somewhat different GPIO. For one thing it is actually two peripherals GPIO and GPIOTE, but I tried to hide that in the API. It also has only one port, so that parameter won't make sense.

**Comment by danielinux** [comment]
Good to go for me. Gpio API generalizations can be completed later on when we fix other targets too.

**Comment by maksymko** [comment]
Thank you for the quick review!

**Comment by brabo** [comment]
@maksymko great job! Good to go ;)
